digits/tools/tensorflow/tf_data.py

```diff
@@ -0,0 +1,893 @@
+            self.keys = [self.db_path]
+            first_file_path = self.db_path
+        elif self.backend == 'filelist':
+            # Single file with a list of files
+            with open(self.db_path) as f:
+                self.keys = f.readlines()
+
+            # Retain only the images in the list
+            self.keys = [key.split(LIST_DELIMITER)[0].rstrip() for key in self.keys]
+
+            if len(self.keys) > 0:
+                # Assume the first entry in the line is a pointer to the file path
+                first_file_path = self.keys[0]
+            else:
+                logging.error('Filelist (%s) contains no lines.' % (self.db_path))
+                exit(-1)
+        else:
+            logging.error('Unsupported backend in FileListLoader (%s)' % (self.backend))
+            exit(-1)
+
+        self.total = len(self.keys)
+
+        # Check first file for statistics
+        im = Image.open(first_file_path)
+        self.width, self.height = im.size
+        self.channels = 1 if im.mode == 'L' else 3  # @TODO(tzaman): allow more channels
+
+        self.data_mime = magic.from_file(first_file_path, mime=True)
+
+        if self.bitdepth == 8:
+            self.image_dtype = tf.uint8
+        else:
+            if self.data_mime == 'image/jpeg':
+                logging.error("Tensorflow does not support 16 bit jpeg decoding.")
+                exit(-1)
+            self.image_dtype = tf.uint16
+
+        self.reader = tf.WholeFileReader()
+
+    def get_queue(self):
+        return tf.train.string_input_producer(
+            self.keys,
+            num_epochs=self.num_epochs,
+            capacity=self.total,
+            shuffle=self.shuffle,
+            seed=self._seed,
+            name='input_producer'
+        )
+
+    def get_single_data(self, key_queue):
+        """
+        Returns:
+            key, single_data, single_data_shape, single_label, single_label_shape
+        """
+        key, value = self.reader.read(key_queue)
+        shape = np.array([self.width, self.height, self.channels], dtype=np.int32)  # @TODO: this is not dynamic
+        return key, value, shape  # @TODO(tzaman) - Note: will only work for inferencing stage!
+
+
+class TFRecordsLoader(LoaderFactory):
+    """ The TFRecordsLoader connects directly into the tensorflow graph.
+    It uses TFRecords, the 'standard' tensorflow data format.
+    """
+    def __init__(self):
+        pass
+
+    def initialize(self):
+        self.float_data = False  # For now only strings
+        self.unencoded_data_format = 'hwc'
+        self.unencoded_channel_scheme = 'rgb'
+        self.reader = None
+        if self.bitdepth == 8:
+            self.image_dtype = tf.uint8
+        else:
+            self.image_dtype = tf.uint16
+
+        # Count all the records @TODO(tzaman): account for shards!
+        # Loop the records in path @TODO(tzaman) get this from a txt?
+        # self.db_path += '/test.tfrecords' # @TODO(tzaman) this is a hack
```

Comment:
I'd leave these lines in for now. The TFRecords support needs some work. Maybe someone wants to pick that up.